firestore/src/main/app_check_credentials_provider_desktop.h

```diff
@@ -0,0 +1,98 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIREBASE_FIRESTORE_SRC_MAIN_APP_CHECK_CREDENTIALS_PROVIDER_DESKTOP_H_
+#define FIREBASE_FIRESTORE_SRC_MAIN_APP_CHECK_CREDENTIALS_PROVIDER_DESKTOP_H_
+
+#include <memory>
+#include <mutex>  // NOLINT(build/c++11)
+#include <string>
+
+#include "Firestore/core/src/credentials/credentials_fwd.h"
+#include "Firestore/core/src/credentials/credentials_provider.h"
+#include "Firestore/core/src/credentials/user.h"
+#include "Firestore/core/src/util/statusor.h"
+#include "app/src/include/firebase/app.h"
+#include "app/src/include/firebase/future.h"
+
+#if defined(__ANDROID__)
+#error "This header should not be used on Android."
+#endif
+
+namespace firebase {
+namespace firestore {
+
+// Glues together C++ Firebase App Check and Firestore: allows Firestore to
+// listen to App Check events and to retrieve App Check tokens. Thread-safe.
+//
+// This is a language-specific implementation of `AppCheckCredentialsProvider`
+// that works with the public C++ App Check.
+class CppAppCheckCredentialsProvider
+    : public firestore::credentials::AppCheckCredentialsProvider {
+ public:
+  explicit CppAppCheckCredentialsProvider(App& app);
+  ~CppAppCheckCredentialsProvider() override;
+
+  CppAppCheckCredentialsProvider(const CppAppCheckCredentialsProvider&) =
+      delete;
+  CppAppCheckCredentialsProvider& operator=(
+      const CppAppCheckCredentialsProvider&) = delete;
+
+  // `firestore::credentials::AppCheckCredentialsProvider` interface.
+  void SetCredentialChangeListener(
+      firestore::credentials::CredentialChangeListener<std::string> listener)
+      override;
+  void GetToken(
+      firestore::credentials::TokenListener<std::string> listener) override;
+
+ private:
+  void AddAppCheckStateListener();
+  void RemoveAppCheckStateListener();
+
+  // Callback for the function registry-based pseudo-AuthStateListener
+  // interface.
+  static void OnAppCheckStateChanged(std::string, void* context);
+
+  // Wraps the data that is used by `auth::User::GetToken` callback. This
+  // credentials provider holds a shared pointer to `Contents`, while the
+  // `GetToken` callback stores a weak pointer. This makes safe the case where
+  // the `GetToken` callback might be invoked after this credentials provider
+  // has already been destroyed (Auth may outlive Firestore).
+  struct Contents {
+    explicit Contents(App& app) : app(app) {}
+
+    // FirebaseCppCredentialsProvider may be used by more than one thread. The
+    // mutex is locked in all public member functions and none of the private
+    // member functions (with the exception of `RequestToken` that locks the
```

Comment:
This bit of the comment can be removed. RequestToken is an auth method and the similar logic is just done inline in the public GetToken method now.

That said, in Auth RequestToken obtains the lock during the callback where it calls the listener with the token. We may want to obtain the lock similarly in the OnCompletion callback in GetToken. If so this comment can be kept but rephrased slightly to mention the GetToken lambda instead of RequestToken lambda.

Reply:
After going over it again, I don't believe it is needed. Nothing in the listener of GetToken requires anything from the object, I simplified to just have the app directly, instead of the additional wrapping around it.

Reply:
Removing the 'contents' object makes sense. I see that you also removed the mutex entirely. I think that is reasonable, as the only variable that I see that might be modified is change_listener_, and it's likely just set up-front and then left alone, compared to auth which has the token_generation field which is frequently modified. So no mutex is probably fine here.